subsys/net/lib/dhcpv4/dhcpv4.c

```diff
@@ -667,6 +671,71 @@ static void dhcpv4_timeout(struct k_work *work)
 	}
 }
 
+#if defined(CONFIG_NET_DHCPV4_OPTION_CALLBACKS_VENDOR_SPECIFIC)
+
+static int dhcpv4_parse_option_vendor(struct net_pkt *pkt, struct net_if *iface,
+				      enum net_dhcpv4_msg_type *msg_type, int length)
+{
+	struct net_dhcpv4_option_callback *cb, *tmp;
+	struct net_pkt_cursor backup;
+	uint8_t len;
+	uint8_t type;
+
+	if (length < 3) {
+		NET_ERR("Vendor-specific option parsing, length too short");
+		net_pkt_skip(pkt, length);
+		return -EBADMSG;
+	}
+
+	while (!net_pkt_read_u8(pkt, &type)) {
+		if (type == DHCPV4_OPTIONS_END) {
+			NET_DBG("Vendor-specific options_end");
+			return 0;
+		}
+		length--;
+
+		if (length <= 0) {
+			NET_ERR("Vendor-specific option parsing, malformed option");
+			return -EBADMSG;
+		}
+
+		if (net_pkt_read_u8(pkt, &len)) {
+			NET_ERR("Vendor-specific option parsing, bad length");
+			return -ENOBUFS;
+		}
+		length--;
+		if (length < len) {
+			NET_ERR("Vendor-specific option parsing, length too long");
+			net_pkt_skip(pkt, length);
```

Comment:
If there is an error, we should discard the message and not try to do something like this. The packet could be formatted by a malicious actor so that it could lead to system compromise. After all, we are here parsing data that is coming from outside of the device and we should be very strict what we are accepting.

Reply:
With the code I added, I want to make sure 
1. that a dhcpv4 packet with vendor specific options, that is not encapsulated, will not break the program 
2.  that a non encapsulated packet will not result in a drop of the whole packet, which would result in not  getting a IPv4-Address, instant it will just ignore the vendor specific option and will go to the next option. 

To make extra sure we could add the cursor backup around the funktion. 
`net_pkt_cursor_backup(pkt, &backup);
				dhcpv4_parse_option_vendor(pkt, iface, msg_type, length);
				net_pkt_cursor_restore(pkt, &backup);`

Reply:
If we can ignore safely the invalid vendor specific data, then that could be ok. I am a bit worried about doing `net_pkt_skip()` here, are we able to sync the parsing properly if there is a length error.

Reply:
I started working on improving the tests to cover the most likely cases of parsing going out of sync (mostly invalid lengths in the encoded data). I'll add them later today

Reply:
The tests now check for short packets in multiple cases, including the edge case of option 43 with a length of only 1. (I.e. only enough data for the option type and not even a length).
Fins code passes all of my tests.